### DISCUSSION TITLE
using Processing libraries in py5

### DISCUSSION THREAD

**Discussion by Maltiec** [py5coding/py5generator#79] [Show and tell]
Hello! First of all many thanks to the author of py5 library! I'm using it instead of classic Processing Python mode in my course on Generative art and it is really helpful. Now, I've tried to use Processing libs, followed the instructions on Camera3D, it worked, but I am using Anaconda-Spyder toolchain, and it is really awful to restart kernel everytime before starting sketch. So I've just moved libraries to my conda env's folder (in my case it is: home/evgeniy/anaconda3/envs/py5coding/lib/python3.8/site-packages/py5/jars/) and it seems to work fine for now. The only trick is that sometimes library name is not the same as the name of .jar file, and it can lead to a problem. For example to use VideoExport library you need to write:

`from com.hamoid import VideoExport`

Hope it will help somebody.)

**Comment by hx2A** [comment]
Welcome, @Maltiec ! I'm very happy to hear you are using py5 in your class. Someday I'd love to see your student's projects. For me it is a thrill to see what people are doing with py5, and I wish more people would show me their work. If you are on twitter, use the hashtag `#py5` or the handle `@py5coding`.

Can I have more information about why you are restarting the kernel before starting a sketch? Which kernel are you using? There are several options available to you for adding jars to use in a py5 Sketch.

First, you can use [py5_tools.add_classpath()](http://py5.ixora.io/reference/py5tools_add_classpath.html) or [py5_tools.add_jars()](http://py5.ixora.io/reference/py5tools_add_jars.html). These must be used before importing py5. This isn't useful if you are using the py5 kernel because it will import py5 for you, before you get a chance to use either of those functions. I suspect this is the issue you are dealing with.

Alternatively, you can put jars in a directory called "jars" that is a subdirectory of the current working directory (`os.getcwd()`). Any jars in there or in a subdirectory of that location will be loaded automatically before importing py5.

You can also add jars to the `CLASSPATH` environment variable, just like you would for any other Java application.

The next version of py5 (due out in May) responds to a `PY5_JARS` environment variable. This will be set to a directory. Any jar in that directory or a subdirectory will be added to the classpath. This will be more useful than the other options as you can make a local repository of Processing jars that you like to use and they will be available for you for all of your projects.

**Reply by Maltiec** [reply]
By kernel I mean Spyder IDE console, so I think it is different kernel. ) Yep I've used py5_tools.add_jars(), and it works when I launch my sketch for the first time. But then, after stopping the sketch, Java machine is still working in the Spyder console. So there are two ways to start sketch without an error: 1. comment the lines with py5_tools. 2. the way I did - just copy everything to jar directory manually. The news about environment variable is really great! )  I've started to post my experiments in Instagram with #py5 , may be will also do so in LinkedIn. I rarely use twitter though.

**Reply by hx2A** [reply]
OK, I see what you are saying. You've already started the JVM and will get an error if you call `py5_tools.add_jars()` a second time. How about structuring your code like this:

``` python
import py5_tools

if not py5_tools.is_jvm_running():
    py5_tools.add_jars(...)

import py5
```

That might help.

I just found you on Instagram. I'm looking forward to seeing what you do!

**Reply by Maltiec** [reply]
yep, it works! I saw you followed me.)
